docs/coding-guidelines/source-generator-guidelines.md

```diff
@@ -0,0 +1,27 @@
+Recommended reading to better understand source generators,
+[Roslyn Source Generators Cookbook](https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md).
+
+[Project guidance](./project-guidelines.md#directory-layout)
+
+[Packaging guidance](./libraries-packaging.md#analyzers--source-generators)
+
+## Source Generator Best Practices
+
+### DOs
+
+- **DO** generate code that looks as if a developer would write it manually.
+- **DO** emit strings rather than using the Roslyn Syntax API for better performance.
+- **DO** use consistent indentation and formatting in generated code.
+- **DO** make generators incremental whenever possible using the [`IIncrementalGenerator`](https://learn.microsoft.com/dotnet/api/microsoft.codeanalysis.iincrementalgenerator) interface.
+- **DO** cache intermediate results to avoid redundant computation.
+- **DO** consider the impact on build time and optimize accordingly.
+- **DO** report clear diagnostics when errors occur during generation.
```

Comment:
The recommended guidance is not to emit diagnostics from the generators, but rather to have generators bypass the logic if they detect an invalid state in the appropriate way (eg. either not generating code, or generating empty stubs just to avoid misleading build errors), and then having a separate analyzer actually produce the diagnostics.